lib/matplotlib/cbook/__init__.py

```diff
@@ -2002,6 +2002,47 @@ def _array_perimeter(arr):
     ))
 
 
+def _unfold(arr, axis, size, step):
+    """Append an extra dimension containing sliding windows along ``axis``.
+
+    All windows are of size ``size`` and begin with every ``step`` elements.
+
+    Parameters
+    ----------
+    arr : ndarray, shape (N_1, ..., N_k)
+        The input array
+
+    Returns
+    -------
+    windows : ndarray, shape (N_1, ..., 1 + (N_axis-size)/step, ..., N_k, size)
+    """
+    new_shape = list(arr.shape) + [size]
+    new_strides = list(arr.strides) + [arr.strides[axis]]
+    new_shape[axis] = (new_shape[axis] - size) // step + 1
+    new_strides[axis] = new_strides[axis] * step
+    return np.lib.stride_tricks.as_strided(arr,
+                                           shape=new_shape,
+                                           strides=new_strides,
+                                           writeable=False)
+
+
+def _array_patch_perimeters(x, rstride, cstride):
+    """Extract perimeters of patches from ``arr``.
+
+    Extracted patches are of size ``(rstride + 1) x (cstride + 1)`` and share
+    perimeters with their neighbors.
+    """
+    assert rstride > 0 and cstride > 0
+    assert (x.shape[0] - 1) % rstride == 0
+    assert (x.shape[1] - 1) % cstride == 0
+    upper = _unfold(x[:-1:rstride, :-1], 1, cstride, cstride)
+    lower = _unfold(x[rstride::rstride, 1:], 1, cstride, cstride)[..., ::-1]
+    right = _unfold(x[:-1, cstride::cstride], 0, rstride, rstride)
+    left = _unfold(x[1:, :-1:cstride], 0, rstride, rstride)[..., ::-1]
+    return (np.concatenate((upper, right, lower, left), axis=2)
+              .reshape(-1, 2 * (rstride + cstride)))
```

Comment:
it may be worth mentioning in the docstring the shape of this array (as you did for _unfold)